### PR TITLE
Adding minor misc fixes for pfcwd scripts.

### PR DESCRIPTION
**Pull request by rraghav-cisco**
### Description of PR
This PR addresses the following issues:
- The dut-counter check on pfcwd scripts is using a wrong logic, it attempts to read the incoming drops and incoming packets and compares. Unfortunately, the packets are not being dropped in ingress, but they are dropped in egress. So this PR changes the logic to read the outgoing port's drops,  and incoming ports' total packets. 
- The flow-completion-check in the common code is for 20 seconds. But some of the traffic streams is configured for 26 seconds, so we need to check for atleast 26 seconds. So I have increased the time to 30 seconds.
- Along with the above, the pfcwd_basic script is updated to use the common code as implemented in the PR:15099.

Summary:
Pls see above.

### Type of change

- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [X] 202405

#### What is the motivation for this PR?
- Fixing some of the issues that we came across during the testing.
- move the repeated code to the common code.
#### How did you do it?
- Updated the counter reading mechanism.
- Updated the max_attempts for flow-completion to 30 seconds.
- Updated the pfcwd_basic to use the new lib code from #15099.
#### How did you verify/test it?
- Ran it on my TB:

#### Any platform specific information?
Changes were tested on cisco-8000 only. The one fail below is not related to this change.

My run result:
`=========================================================================================================================== PASSES ===========================================================================================================================
______________________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio[multidut_port_info0-True] _______________________________________________________________________________________________
______________________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio[multidut_port_info1-True] _______________________________________________________________________________________________
______________________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio[multidut_port_info1-False] ______________________________________________________________________________________________
----------------------------------------------------------------------------- generated xml file: /run_logs/ixia/fixturize-reboot/2024-10-30-21-43-12/tr_2024-10-30-21-43-12.xml -----------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
21:54:15 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio[multidut_port_info0-True]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio[multidut_port_info1-True]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio[multidut_port_info1-False]
FAILED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio[multidut_port_info0-False] - Failed: Loss rate of Data Flow 1 (0.17994320225890004) should be in [0, 0]
==================================================================================================== 1 failed, 3 passed, 7 warnings in 660.78s (0:11:00) =====================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$ 
`

### CODE REVIEW

**Comment by rraghav-cisco** [comment]
@sdszhang , @selldinesh , @kamalsahu0001 , @amitpawar12 : For your comments.

This PR can be merged only after #15099  is merged.

**Review by zhixzhu — APPROVED** [review]

**Comment by rraghav-cisco** [comment]
> @rraghav-cisco this PR failed at test validation. can you fix it?

@sdszhang , sorry for that. I fixed the imports.

**Comment by sdszhang** [comment]
@amitpawar12 can you review this one? Thanks.

File: tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py
Comment (sdszhang) [inline tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py:135]: can we add the 'rx_fail' field in get_interface_stats instead of using creating a new function?

Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfc/files/lossless_response_to_external_pause_storms_helper.py:135]: Done. The new result after change:
```
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info0-yy39top-lc4|0]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info0-yy39top-lc4|1]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info0-yy39top-lc4|2]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info0-yy39top-lc4|5]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info0-yy39top-lc4|6]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info1-yy39top-lc4|0]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info1-yy39top-lc4|1]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info1-yy39top-lc4|2]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info1-yy39top-lc4|5]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio[multidut_port_info1-yy39top-lc4|6]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_multi_lossy_prio[multidut_port_info0]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_multi_lossy_prio[multidut_port_info1]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info0-cold-yy39top-lc4|0]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info0-cold-yy39top-lc4|1]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info0-cold-yy39top-lc4|2]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info0-cold-yy39top-lc4|5]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info0-cold-yy39top-lc4|6]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info1-cold-yy39top-lc4|0]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info1-cold-yy39top-lc4|1]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info1-cold-yy39top-lc4|2]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info1-cold-yy39top-lc4|5]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_single_lossy_prio_reboot[multidut_port_info1-cold-yy39top-lc4|6]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_multi_lossy_prio_reboot[multidut_port_info0-cold]
PASSED snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py::test_pfc_pause_multi_lossy_prio_reboot[multidut_port_info1-cold]
SKIPPED [10] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:135: Reboot type warm is not supported on cisco-8000 switches
SKIPPED [10] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:135: Reboot type fast is not supported on cisco-8000 switches
SKIPPED [2] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:195: Reboot type warm is not supported on cisco-8000 switches
SKIPPED [2] snappi_tests/multidut/pfc/test_multidut_pfc_pause_lossy_with_snappi.py:195: Reboot type fast is not supported on cisco-8000 switches
================================================================================================= 24 passed, 24 skipped, 28 warnings in 20114.16s (5:35:14) ==================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$
```

**Review by sdszhang — APPROVED** [review]
LGTM

**Comment by rraghav-cisco** [comment]
Result after recent fix:
```
=========================================================================================================================== PASSES ===========================================================================================================================
_________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio_reboot[multidut_port_info0-cold-yy39top-lc4|4-True] __________________________________________________________________________________
_________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio_reboot[multidut_port_info1-cold-yy39top-lc4|4-True] __________________________________________________________________________________
--------------------------------------------------------------------------------- generated xml file: /run_logs/ixia/old_func/2024-11-10-16-15-48/tr_2024-11-10-16-15-48.xml ---------------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
16:59:51 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio_reboot[multidut_port_info0-cold-yy39top-lc4|4-True]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio_reboot[multidut_port_info1-cold-yy39top-lc4|4-True]
SKIPPED [2] snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:129: Reboot type warm is not supported on cisco-8000 switches
SKIPPED [2] snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:129: Reboot type fast is not supported on cisco-8000 switches
=================================================================================================== 2 passed, 4 skipped, 6 warnings in 2641.75s (0:44:01) ====================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$ 
```

File: tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py
Comment (sdszhang) [inline tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py:144]: why 1024 packet size doesn't work? I would prefer to leave this change in a separate PR.

Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py:144]: For long-link+HBM, Alpesh recommended to use 8K packet size, since that is the buffer size. I have not verified it on other platforms, but my deduction is it should not affect all other cisco platforms also.

Comment (sdszhang) [reply tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py:144]: can you remove this last commit and put in another PR? as it's cisco platform specific.  Then, we can merge in the rest of the change first.    We can continue discuss the platform limitation in the new PR.

@abdosi for viz too.

Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfcwd/files/pfcwd_multidut_basic_helper.py]: Done. Will raise a different PR for the cisco-8000 changes.

**Pull request merged**
